runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -586,52 +689,119 @@ bool J9::ValuePropagation::transformUnsafeCopyMemoryCall(TR::Node *arraycopyNode
          TR::Node *destOffset = arraycopyNode->getChild(4);
          TR::Node *len        = arraycopyNode->getChild(5);
 
-         bool isGlobal;
-         TR::VPConstraint *srcOffsetConstraint = getConstraint(srcOffset, isGlobal);
-         TR::VPConstraint *dstOffsetConstraint = getConstraint(destOffset, isGlobal);
-         TR::VPConstraint *copyLenConstraint   = getConstraint(len, isGlobal);
+#if defined(J9VM_GC_ENABLE_SPARSE_HEAP_ALLOCATION)
+         if (TR::Compiler->om.isOffHeapAllocationEnabled())
+            {
+            // When using balanced GC policy with offheap allocation enabled, there are three possible for an argument type:
+            //
+            // 1.) The type is known to be a non-array object at compile time. In this scenario, the final address
+            //     can be calculated by simply adding ref and offset.
+            // 2.) The type is known to be an array at compile time. In this scenario, if the ref at runtime is `null` then
+            //     final address is calculated as in case 1. If not `null` then final address is the adjusted ref, by loading 
+            //     the dataAddr pointer field then add it to the offset.
+            // 3.) The type of the object at dest is unknown at compile time (type is `java/lang/Object`). In this scenario, 
+            //     a runtime null check and arrayCHK must be generated to determine whether it needs to be handled such as
+            //     case 1 or case 2.
+
+            // Check src/dest type at compile time
+            int srcSigLen, destSigLen;
+            int objSigLength = strlen("Ljava/lang/Object;");
+            const char *srcObjTypeSig = src->getSymbolReference() ? src->getSymbolReference()->getTypeSignature(srcSigLen) : 0;
+            const char *destObjTypeSig = dest->getSymbolReference() ? dest->getSymbolReference()->getTypeSignature(destSigLen) : 0;
+            // Case 3
+            bool srcArrayCheckNeeded = srcObjTypeSig == NULL || strncmp(srcObjTypeSig, "Ljava/lang/Object;", objSigLength) == 0;
+            bool destArrayCheckNeeded = destObjTypeSig == NULL || strncmp(destObjTypeSig, "Ljava/lang/Object;", objSigLength) == 0;
+            // Case 2 & 3
+            bool srcAdjustmentNeeded = srcArrayCheckNeeded || srcObjTypeSig[0] == '[';
+            bool destAdjustmentNeeded = destArrayCheckNeeded || destObjTypeSig[0] == '[';
```

Comment:
After discussions it seems that using constraints will be delayed as an after enabled improvement. More types will be added to the signature check, and if signature is not available it will fallback to the conservative path. Checking with VM about interfaces carrying array objects should cause Abstract Method Error.

Reply:
> Checking with VM about interfaces carrying array objects should cause Abstract Method Error.

What do you mean?

Reply:
"We might not need to be concerned with unverified handwritten bytecodes (notice javac doesn’t allow `ArbitraryInterfaceType myObject = new SomeType[];`).  That sort of bytecode will result in J9 throwing AME (Abstract Method Error)."

Reply:
I don't think this kind of thing can cause `AbstractMethodError`. The most typical consequence of passing an arbitrary object where the signature suggests a particular interface is expected would be `IncompatibleClassChangeError`, which is what `invokeinterface` throws when the receiver doesn't implement the interface. But that will only happen if the object actually reaches an incompatible `invokeinterface` instruction that uses it as the receiver

I'll grant that it's a rare and strange situation, but (in the absence of any other problem) bytecode that ignores interface types is perfectly valid bytecode. A method could pass a (supposed) `Runnable` into `copyMemory()` and at runtime that `Runnable` could turn out to be any object whatsoever, including an array

Since you're not yet using VP constraints, and you won't be able to use them here in `TransformUtil` anyway, I suggest calling `getClassFromSignature()` and making sure you get back a (non-null) non-interface class. If you know that it's a specific non-interface class, then you know that the signature can be trusted. But if it's an interface (or unknown), then it can't be relied upon

Reply:
Updated signature check to check for interface.